### PR TITLE
Add Last Turn Annotation Only option to turn annotations static task

### PR DESCRIPTION
**Pull request by mwillwork**
**Patch description**

**Testing steps**

**Other information**

### CODE REVIEW

File: parlai/crowdsourcing/tasks/turn_annotations_static/conf/example.yaml [hunk 1]
```diff
@@ -1,5 +1,9 @@
 #@package _global_
 mephisto:
+  architect:
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/turn_annotations_static/conf/example.yaml:3]: Hmm - it'd be good to keep these lines in the `run.py` file for now, for consistency with the other Mephisto tasks. I've made a note to hopefully move these lines into the `example.yaml` files of *all* tasks once we move to Hydra 1.1

File: parlai/crowdsourcing/tasks/turn_annotations_static/run.py [hunk 2]
```diff
@@ -21,13 +21,12 @@
 
 TASK_DIRECTORY = os.path.dirname(os.path.abspath(__file__))
 
+# To run the task with your own config outside this folder (recommended!)
+# Use the command formulation below. For more info,
+# check the README in parlai/crowdsourcing/
+# python turn_annotations_static/run.py conf=<conf name sans yaml> --config-dir <path to directory with a conf/ folder>
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/turn_annotations_static/run.py:27]: Hmm - in general I'm trying not to repeat the same documentation in both `run.py` and the README, because if we change the command for how to run this later on (which we will), we might forget to update the documentation in both places. Lemme know if you feel strongly about this, and we can discuss further

File: parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py [hunk 3]
```diff
@@ -60,6 +60,12 @@ class TurnAnnotationsStaticBlueprintArgs(StaticReactBlueprintArgs):
             "help": "Specify which utterance indices to annotate per conversation in a JSONL file. Must be same length as conversations data-jsonl file. See example file in task_config/annotation_indices_example.jsonl"
         },
     )
+    annotation_last_only: Optional[bool] = field(
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py:63]: nit: maybe "annotate_last_only" is cleaner here?

File: parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py [hunk 4]
```diff
@@ -125,6 +132,10 @@ def __init__(
         # conversation to annotate
         self.annotation_indices = None
         if self.args.blueprint.annotation_indices_jsonl:
+            if self.annotation_last_only:
+                raise RuntimeError(
+                    'Cannot use flag annotation_last_only an supply a file with annotation indices.'
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py:137]: nit: typo in "and"

File: parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py [hunk 5]
```diff
@@ -215,15 +226,28 @@ def process_data(self, data_dicts, annotation_indices=None):
                     )
                 for a in annotation_indices[conv_idx]:
                     processed_dialog = self._process_conversation(d, [a])
+                    total_annotation_count += 1
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py:229]: hmm - won't this double-count given the `total_annotation_count += len(annotation_indices[conv_idx])` line above?

File: parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py [hunk 6]
```diff
@@ -215,15 +226,28 @@ def process_data(self, data_dicts, annotation_indices=None):
...
+                # Annotate only the last utterance
+                last_idx = len(d['dialog']) * 2 - 1
+                processed_dialog = self._process_conversation(
+                    d, annotation_indices=[last_idx]
+                )
+                total_annotation_count += 1
+                output.append(processed_dialog)
             else:
+                # Annotate all the model utterances
+                total_annotation_count += len(d['dialog']) / 2
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py:241]: Yes, good catch

File: parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py [hunk 7]
```diff
@@ -215,15 +226,28 @@ def process_data(self, data_dicts, annotation_indices=None):
...
                 processed_dialog = self._process_conversation(
                     d, annotation_indices=None
                 )
                 output.append(processed_dialog)
+
         print(
-            f'Processed {len(data_dicts)} total conversations into {len(output)} conversations to be used in crowdsourcing task with {total_annotation_count} total annotations.'
+            f'process_data: Processed {len(data_dicts)} total conversations into {len(output)} conversations in the full data with {total_annotation_count} total turn annotations. (Does not account for units per assignment values.)'
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py:248]: Hmm can you clarify "(Does not account for units per assignment values.)" further?

File: parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py [hunk 8]
```diff
@@ -257,7 +281,8 @@ def _process_conversation(self, d, annotation_indices: Optional[List[int]] = Non
                 continue
             # If there is a persona, which is context as in the ConvAI2
             # task, we don't want to display the persona utterances
-            if 'persona' not in full_turn[0]['text']:
+            # Dataset has utterance that has "your personal"!
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py:284]: Nit: maybe reword to `Persona strings have the format "your persona:"`

File: parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py [hunk 9]
```diff
@@ -287,6 +312,10 @@ def _process_conversation(self, d, annotation_indices: Optional[List[int]] = Non
             raise Exception(
                 f'Conversation had {adjusted_turn_idx} but max_turn_to_show was {max_turn_to_show}'
             )
+        assert np.any(
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/turn_annotations_static/turn_annotations_blueprint.py:315]: nit: `any(nd['do_annotate'] for nd in new_dialogue)` would be simpler here

**Review by EricMichaelSmith — CHANGES_REQUESTED** [review]
Great to have this feature! I see that the crowdsourcing checks have failed, perhaps due to having changed the `defaults` list in `run.py`? I have some other comments, many of which are nits

File: parlai/crowdsourcing/tasks/turn_annotations_static/run.py (hunk 2 above)
Comment (mwillwork) [reply parlai/crowdsourcing/tasks/turn_annotations_static/run.py:27]: I think I had mentioned to you, but I think we should remove the README. When you open the run.py, you expect tot have instructions there.

**Review by EricMichaelSmith — APPROVED** [review]
Thanks for making those changes! I just made a fix that will hopefully get the crowdsourcing CI checks to work again so we can merge this in
